src/System.Collections/src/System/Collections/Generic/HashSetEqualityComparer.cs

```diff
@@ -20,26 +15,68 @@ internal sealed class HashSetEqualityComparer<T> : IEqualityComparer<HashSet<T>>
 
         public HashSetEqualityComparer()
         {
-            _comparer = EqualityComparer<T>.Default;
+            // null comparer signals legacy use.
         }
 
-        // using m_comparer to keep equals properties in tact; don't want to choose one of the comparers
-        public bool Equals(HashSet<T> x, HashSet<T> y)
+        public HashSetEqualityComparer(IEqualityComparer<T> comparer)
         {
-            return HashSet<T>.HashSetEquals(x, y, _comparer);
+            _comparer = comparer ?? EqualityComparer<T>.Default;
         }
 
+        public bool Equals(HashSet<T> x, HashSet<T> y)
+            => _comparer == null ? HashSet<T>.LegacyHashSetEquals(x, y) : HashSet<T>.HashSetEquals(x, y, _comparer);
+
         public int GetHashCode(HashSet<T> obj)
         {
-            int hashCode = 0;
-            if (obj != null)
+            if (obj == null)
             {
+                return 0;
+            }
+
+            if (_comparer == null)
+            {
+                IEqualityComparer<T> comparer = EqualityComparer<T>.Default;
+                // Legacy behaviour for compat.
+                int hashCode = 0;
                 foreach (T t in obj)
                 {
-                    hashCode = hashCode ^ (_comparer.GetHashCode(t) & 0x7FFFFFFF);
+                    hashCode = hashCode ^ (comparer.GetHashCode(t) & 0x7FFFFFFF);
                 }
-            } // else returns hashcode of 0 for null hashsets
-            return hashCode;
+
+                return hashCode;
+            }
+
+            unchecked
+            {
+                int hashCode = 1592614637;
+                IEqualityComparer<T> comparer = _comparer;
+
+                // Already know the set is distinct as considered by comparer
```

Comment:
Sounds like you're making a statement here, rather than saying that if this is true, we can avoid some costs.